core/capabilities/compute/compute.go

```diff
@@ -97,35 +105,96 @@ func generateID(binary []byte) string {
 	return fmt.Sprintf("%x", id)
 }
 
-func copyRequest(req capabilities.CapabilityRequest) capabilities.CapabilityRequest {
-	return capabilities.CapabilityRequest{
-		Metadata: req.Metadata,
-		Inputs:   req.Inputs.CopyMap(),
-		Config:   req.Config.CopyMap(),
+func (c *Compute) Execute(ctx context.Context, request capabilities.CapabilityRequest) (capabilities.CapabilityResponse, error) {
+	ch, err := c.enqueueRequest(ctx, request)
+	if err != nil {
+		return capabilities.CapabilityResponse{}, err
+	}
+
+	select {
+	case <-c.stopCh:
+		return capabilities.CapabilityResponse{}, errors.New("service shutting down, aborting request")
+	case <-ctx.Done():
+		return capabilities.CapabilityResponse{}, fmt.Errorf("request cancelled by upstream: %w", ctx.Err())
+	case resp := <-ch:
+		return resp.resp, resp.err
 	}
 }
 
-func (c *Compute) Execute(ctx context.Context, request capabilities.CapabilityRequest) (capabilities.CapabilityResponse, error) {
-	copied := copyRequest(request)
+type request struct {
+	ch  chan response
+	req capabilities.CapabilityRequest
+	ctx func() context.Context
+}
+
+type response struct {
+	resp capabilities.CapabilityResponse
+	err  error
+}
+
+func (c *Compute) enqueueRequest(ctx context.Context, req capabilities.CapabilityRequest) (<-chan response, error) {
+	ch := make(chan response)
+	r := request{
+		ch:  ch,
+		req: req,
+		ctx: func() context.Context { return ctx },
+	}
+	select {
+	case <-c.stopCh:
+		return nil, errors.New("service shutting down, aborting request")
+	case <-ctx.Done():
+		return nil, fmt.Errorf("could not enqueue request: %w", ctx.Err())
+	case c.queue <- r:
+		return ch, nil
+	}
+}
+
+func shallowCopy(m *values.Map) *values.Map {
+	to := values.EmptyMap()
+
+	for k, v := range m.Underlying {
+		to.Underlying[k] = v
+	}
+
+	return to
+}
 
-	cfg, err := c.transformer.Transform(copied.Config)
+func (c *Compute) execute(ctx context.Context, respCh chan response, req capabilities.CapabilityRequest) {
+	// Shallow copy the request.
+	// This is because we mutate its overall shape.
+	req = capabilities.CapabilityRequest{
+		Config: shallowCopy(req.Config),
+
+		// These aren't mutated so we ignore them.
+		Metadata: req.Metadata,
+		Inputs:   req.Inputs,
+	}
+
+	cfg, err := c.transformer.Transform(req.Config)
```

Comment:
I wonder if the transformer should be the one doing the copy, we're tightly coupled to its implementation details by knowing the shallow copy is safe.  Alternatively, you can add a comment to Transfrom saying what it's allowed to do.  I'm leaning towards the first.